presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkDiskPageInput.java

```diff
@@ -0,0 +1,186 @@
+            PrestoSparkBroadcastTableCacheManager prestoSparkBroadcastTableCacheManager,
+            StageId stageId,
+            PlanNodeId planNodeId,
+            int storageBasedBroadcastJoinReadBufferSize)
+    {
+        this.pagesSerde = requireNonNull(pagesSerde, "pagesSerde is null");
+        this.tempStorage = requireNonNull(tempStorage, "tempStorage is null");
+        this.broadcastTableFilesInfo = requireNonNull(broadcastTableFilesInfo, "broadcastTableFilesInfo is null");
+        this.tempDataOperationContext = requireNonNull(tempDataOperationContext, "tempDataOperationContext is null");
+        this.prestoSparkBroadcastTableCacheManager = requireNonNull(prestoSparkBroadcastTableCacheManager, "prestoSparkBroadcastTableCacheManager is null");
+        this.stageId = requireNonNull(stageId, "stageId is null");
+        this.planNodeId = requireNonNull(planNodeId, "planNodeId is null");
+        this.storageBasedBroadcastJoinReadBufferSize = storageBasedBroadcastJoinReadBufferSize;
+    }
+
+    private List<Iterator<Page>> getPageIterators()
+    {
+        if (pageIterators == null) {
+            pageIterators = getPages(broadcastTableFilesInfo, tempStorage, tempDataOperationContext, prestoSparkBroadcastTableCacheManager, stageId, planNodeId);
+        }
+        return pageIterators;
+    }
+
+    private List<Iterator<Page>> getPages(
+            List<List<PrestoSparkStorageHandle>> broadcastTableFilesInfo,
+            TempStorage tempStorage,
+            TempDataOperationContext tempDataOperationContext,
+            PrestoSparkBroadcastTableCacheManager prestoSparkBroadcastTableCacheManager,
+            StageId stageId,
+            PlanNodeId planNodeId)
+    {
+        // Try to get table from cache
+        List<List<Page>> pages = prestoSparkBroadcastTableCacheManager.getCachedBroadcastTable(stageId, planNodeId);
+        if (pages == null) {
+            pages = broadcastTableFilesInfo.stream()
+                    .map(tableFiles -> {
+                        List<SerializedPage> serializedPages = loadBroadcastTable(tableFiles, tempStorage, tempDataOperationContext);
+                        return serializedPages.stream().map(serializedPage -> pagesSerde.deserialize(serializedPage)).collect(Collectors.toList());
+                    }).collect(Collectors.toList());
+
+            // Cache deserialized pages
+            prestoSparkBroadcastTableCacheManager.cache(stageId, planNodeId, pages);
+        }
+
+        return pages.stream().map(pageList -> pageList.iterator()).collect(Collectors.toList());
+    }
+
+    private List<SerializedPage> loadBroadcastTable(
+            List<PrestoSparkStorageHandle> broadcastTaskFilesInfo,
+            TempStorage tempStorage,
+            TempDataOperationContext tempDataOperationContext)
+    {
+        try {
+            CRC32 checksum = new CRC32();
+            List<SerializedPage> pages = new ArrayList<>();
+            shuffle(broadcastTaskFilesInfo);
```

Comment:
This could potentially be immutable or shared, I would recommend creating a copy, and then doing shuffle